src/orion/client/runner.py

```diff
@@ -0,0 +1,323 @@
+
+            # Scatter the new trials to our free workers
+            with self.stat.time("scatter"):
+                self.scatter(new_trials)
+
+            # Gather the results of the workers that have finished
+            with self.stat.time("gather"):
+                self.gather()
+
+            if self.is_idle:
+                idle_end = time.time()
+                idle_time += idle_end - idle_start
+                idle_start = idle_end
+
+                log.debug(f"Workers have been idle for {idle_time:.2f} s")
+            else:
+                idle_start = time.time()
+                idle_time = 0
+
+            if self.is_idle and idle_time > self.idle_timeout:
+                msg = f"Workers have been idle for {idle_time:.2f} s"
+
+                if self.pending_trials:
+                    msg = (
+                        f"{msg}; but not all trials finished; "
+                        f"pending_trials: {self.pending_trials}"
+                    )
+
+                if self.has_remaining and not self.is_done:
+                    msg = (
+                        f"{msg}; worker has leg room (has_remaining: {self.has_remaining})"
+                        f" and optimization is not done (is_done: {self.is_done})"
+                    )
+
+                raise LazyWorkers(msg)
+
+        return self.trials
+
+    def sample(self):
+        """Sample new trials for all free workers"""
+        ntrials = len(self.pending_trials) + self.trials
+        remains = self.max_trials_per_worker - ntrials
+
+        # try to get more work
+        new_trials = []
+        n_trial = min(self.free_worker, remains)
+        should_sample_more = (not self.is_done) and self.free_worker > 0 and remains > 0
+
+        if should_sample_more:
+            # the producer does the job of limiting the number of new trials
+            # already no need to worry about it
+            # NB: suggest reserve the trial already
+            new_trials = self._suggest_trials(n_trial)
+            log.debug(f"Sampled {len(new_trials)} new configs")
+
+        return new_trials
+
+    def scatter(self, new_trials):
+        """Schedule new trials to be computed"""
+        new_futures = []
+        for trial in new_trials:
+            future = self.client.executor.submit(
+                _optimize, trial, self.fct, self.trial_arg, **self.kwargs
+            )
+            self.pending_trials[future] = trial
+            new_futures.append(future)
+
+        self.free_worker -= len(new_futures)
+        self.futures.extend(new_futures)
+        log.debug("Scheduled new trials")
+
+    def gather(self):
+        """Gather the results from each worker asynchronously"""
+        results = []
+        try:
+            results = self.client.executor.async_get(
+                self.futures, timeout=self.gather_timeout
+            )
+        except (KeyboardInterrupt, InvalidResult):
+            self.release_all()
```

Comment:
Where would InvalidResult come from? We don't want to release the trials and set them to interrupted if they are actually broken.

Reply:
Looks like we would be missing tests for this part.

Reply:
Closing this comment in favor of https://github.com/Epistimio/orion/pull/684/files#r780410397